### ISSUE TITLE
initialTaskData is not being set after onboarding

### ISSUE THREAD

**Issue by mjkanji** [facebookresearch/Mephisto#941]
I'm trying to add on-boarding to my task and having some issues. I'm probably just doing something wrong, but any help would be highly appreciated.

When I run the task without onboarding (setting `onboarding_qualification: None` in the config file), the task is working as expected. In order to add onboarding, I'm basically following the various examples and adding the following to my script (the addition are highlighted in comments)

```python
def handle_onboarding(onboarding_data):
    return True

@task_script(default_config_file="local_testing_with_onboarding.yaml")
def main(operator: Operator, cfg: DictConfig) -> None:

    # Added for onboarding
    shared_state = SharedStaticTaskState(
        validate_onboarding=handle_onboarding,
    )

    task_dir = cfg.task_dir
    build_custom_bundle(
        task_dir,
        force_rebuild=cfg.mephisto.task.force_rebuild,
        post_install_script=cfg.mephisto.task.post_install_script,
    )

    # For launching without onboarding
    # operator.launch_task_run(cfg.mephisto)

    # Added for onboarding
    operator.launch_task_run(cfg.mephisto, shared_state)
    operator.wait_for_runs_then_shutdown(skip_input=True, log_rate=30)
```

When I then run the task, the onboarding UI starts up as expected, but after I submit, `initialTaskData` is not populated with the task's data (I've confirmed this with `console.log(initialTaskData)` in the React app) and I need to reload the page. Then, the task UI is populated as expected.

Could you help me figure out what I might be doing wrong and how I can go about debugging the task?

**Comment by JackUrb** [comment]
Hi @mjkanji, this is definitely strange, thanks for surfacing! Often when we've seen this in the past it's during a transition from `initialTaskData` containing the onboarding data to it containing the in-task data, wherein an interface partially renders with the wrong information and things crash. This was lightly tracked in a project note, with the given solution:
```
I think the best route is to have initialTaskData alias two different forms of data under the hood. The thinking here is that onboarding data and in-task data will always be surfaced differently, and then we can key which one is in initialTaskData by the current isOnboarding status.
```
We hadn't implemented this though, as for the most part the resolution is just to ensure the task data is formatted correctly before trying to render.

From the sound of it, though, you're having the task stall out on onboarding (not break), is that correct?

**Comment by mjkanji** [comment]
Hi @JackUrb ! The problem is what you described: `initialTaskData` is not switching from `onboarding_data` to the task data.

I was previously passing the onboarding data as part of my `frontend_args` (and so `onboarding_data` in `shared_state` was an empty list). To fix this, I made the following change to the code snippet above (plus the change needed in the webapp):

```python
    with open(cfg.mephisto.blueprint.onboarding_data, "r", encoding="utf-8-sig") as f:
        shared_state = SharedStaticTaskState(
            onboarding_data=json.loads(f.read()),
            validate_onboarding=handle_onboarding,
        )
```

Anyways, with that change, I've recorded a video showing how the `initialTaskData` remains stuck on the `onboarding_data` unless I refresh the page (you may need to view in full-screen mode to see the text in the browser console):

https://app.usebubbles.com/pvdqfnZ795niFYrzQcBNsm/mephisto-bug

Note that at the 0:44 mark, I've logged to the console whether the onboarding task or the main task is being served, as well as the `initialTaskData` and you can see that in both cases, it's the same object.

But after I refresh the page, at the 1:00 mark, you can see that `initialTaskData` is now an array, which is the expected format.

So, as you said, the task is stalling because the data is not being updated. I'm not sure I understand how I'd implement the suggested fix you've outlined above (and if it's something that can be done on the user's end or if it requires changing Mephisto internals -- i.e., do I need to change my code or the Mephisto code)?

**Comment by mjkanji** [comment]
Hi @JackUrb Any thoughts on how I can proceed with debugging this?

**Comment by JackUrb** [comment]
Hi @mjkanji - sorry for the delay here, things have been a bit crazy.

As I was getting at, the task isn't "stalling" (simply staying in the onboarding state), it's breaking when switching to the full task view. 

You can fix this by, instead of just passing the task data object over to your front end view when not in onboarding, first check to see if that attribute it throws because it is missing is available, and if not render "loading". The full task data will arrive shortly, but if the react app breaks then it can't receive it.

**Comment by mjkanji** [comment]
@JackUrb Thank you so much for your reply! It solved my problem! 🙇‍♂️ 

For anyone else who might be confused by what the above meant (like I was), in layman terms, when switching from onboarding to the main task, there is apparently a bit of a delay such that `isOnboarding` from `useMephistoTask()` is set to False and so the front-end starts loading the `MainTaskComponenent`, but `initialTaskData` is still pointing to the onboarding data, not not the data for the main task, and so all your downstream logic breaks.

To prevent the app from 'breaking', just give it some time to refresh the state and show a loading screen in the meantime.

For example, all I needed to do to solve this was add the following to the start of my `MainTaskComponent`:

```javascript
if (initialTaskData == undefined || !Array.isArray(initialTaskData)) {
    return (
      <div>
        <p> Loading...</p>
      </div>
    );
  }
```

In my case, the onboarding data is an object, while the main task's data is an array so I'm just checking for that. You could change the condition to match whatever your app requires (e.g., by explicitly adding a `dataType` key in the `initialTaskData` object itself and checking this explicitly).

```javascript
initialTaskData = {
  dataType: "onboarding" # or main
  data: ...
}
```
